### PR TITLE
[IMP] l10n_it_edi: DatiFattureCollegate must include all reconciled moves

### PR DESCRIPTION
**Pull request by lordkrandel**
The `move.reversed_move_id` field only represents the credit note that the user has manually created to revert a move, entirely or partially, but it doesn't take into consideration all moves reconciled through other means.

We now use the method `move._get_reconciled_invoices()` which goes through all the partials to retrieve them instead.

This enables also the reconciliation with the outstanding credits case to be covered.

Closed 16.0 PR: odoo/odoo#119132

Task [link](https://www.odoo.com/odoo/project/967/tasks/3281816)
task-3281816

### CODE REVIEW

**Comment by tsb-odoo** [comment]
@jco-odoo LGTM

**Review by epictete — APPROVED** [review]
@robodoo r+

**Comment by jco-odoo** [comment]
@robodoo r-

**Comment by jco-odoo** [comment]
Hey, I wanted to discuss...

**Comment by jco-odoo** [comment]
@tsb-odoo Can not we do a combination of both at least?  In most localizations now, the reversed entry is used and if we change it here, we should kind of be consistent, because Italy does not differ from the rest of the world.

To me, the reconciliation process is something else than initially doing the reverse, but it is true that the reconciliation could provide more information, but it could also be that it is only done afterwards and that the reversed entry gives you more info.  

@epictete Sorry for the rogue r-.  (in the worst case, it can still go on, but please consider)

**Comment by tsb-odoo** [comment]
@jco-odoo Italy does not differ from the rest of the world. We will have to adapt everywhere, progressively. In most cases, the credit note, when posted, will be reconciled automatically with the invoice it comes from. So, be it via this reversed_entry_id or via the _get_reconciled_invoices method, the outcome will be the same. 
While if the user explicitly unreconciled invoice and credit note, then we know he went against the initial symptom (credit note fortuitously triggered from a specific invoice but not reconciled with it anymore. 
So, to me, the PR is valid.

**Comment by jco-odoo** [comment]
@tsb-odoo But invoices might be sent at a time when the reconciliation has not been done yet.  So why not fallback?  The way it is written here, we should remove the reversed_entry_id completely.

**Comment by tsb-odoo** [comment]
> @tsb-odoo But invoices might be sent at a time when the reconciliation has not been done yet. So why not fallback? The way it is written here, we should remove the reversed_entry_id completely.

No need to fallback. The credit note is not rejected. DatiFattureCollegate is optional.
Fallback would mean the user undergoes something he might not want. If he explictly unreconciled invoice and credit note, why would we ignore his choice? 
He can solve it by reconciling again, can't he?

**Comment by lordkrandel** [comment]
The italian use case by law is this: credit notes must have links to invoices they relate to. If theres no reconciliation between the two, you shouldnt send credit notes to the state. If you dont wanna reconcile, or couldnt yet, it s your responsability about sending incomplete data.
This is particularly true and I bet enforced for Public Administration money, for traceability reasons.
So this thing about "but there is no reconciliation yet" to me is nonsense. Am I missing something?

**Comment by tsb-odoo** [comment]
> The italian use case by law is this: credit notes must have links to invoices they relate to. If theres no reconciliation between the two, you shouldnt send credit notes to the state. If you dont wanna reconcile, or couldnt yet, it s your responsability about sending incomplete data. This is particularly true and I bet enforced for Public Administration money, for traceability reasons. So this thing about "but there is no reconciliation yet" to me is nonsense. Am I missing something?

There are many things by law that depend on the user. By law, when you sell alcohol, you must apply the standard rate. If you don't, Odoo doesn't prevent you to apply any other rate, or no rate at all. The information is optional. And the user has the possibility to respect the law principle by reconciling properly. 
Again, the change here brings permissivity without affecting the vast majority of habitual cases. When you create a Credit Note from a specific invoice and post it, Odoo automatically reconciles invoice and credit note. This being said, in the vast majority of cases, the present change is seamless.

**Comment by jco-odoo** [comment]
@tsb-odoo Already: why not provide the existing behaviour as a backup/fallback in stable?

**Comment by tsb-odoo** [comment]
> Fallback would mean the user undergoes something he might not want. If he explictly unreconciled invoice and credit note, why would we ignore his choice?

Already answered above : 
Fallback would mean the user undergoes something he might not want. If he explictly unreconciled invoice and credit note, why would we ignore his choice?

File: addons/l10n_it_edi/data/invoice_it_template.xml [hunk 1]
```diff
@@ -156,10 +156,24 @@
...
+                    <IdDocumento t-out="format_alphanumeric(record.reversed_entry_id.name, -20)"/>
+                    <Data t-out="format_date(record.reversed_entry_id.invoice_date)"/>
+                </DatiFattureCollegate>
+            </t>
+            <t t-elif="record.reversal_move_id">
+                <DatiFattureCollegate>
+                    <IdDocumento t-out="format_alphanumeric(record.reversal_move_id.name, -20)"/>
+                    <Data t-out="format_date(record.reversal_move_id.invoice_date)"/>
+                </DatiFattureCollegate>
+            </t>
```
Comment (jco-odoo) [inline addons/l10n_it_edi/data/invoice_it_template.xml:176]:
```suggestion
```

File: addons/l10n_it_edi/tests/test_edi_export.py [hunk 2]
```diff
@@ -248,6 +248,40 @@ def test_invoice_non_domestic_simplified(self):
...
+                    'price_unit': 800.40,
+                    'tax_ids': [Command.set(purchase_tax.ids)],
+                })
+            ]
+        }
+
+        bill = Move.create({'move_type': 'in_invoice', **values})
+        credit_note = Move.create({'move_type': 'in_refund', **values})
+        (bill + credit_note).action_post()
+        bill.reversed_entry_id = credit_note
```
Comment (jco-odoo) [inline addons/l10n_it_edi/tests/test_edi_export.py:282]:
```suggestion
        credit_note.reversed_entry_id = bill
```

**Comment by jco-odoo** [comment]
@robodoo r+ rebase-merge

**Pull request closed without merging**